### PR TITLE
fix(server): fix logpath causing Server.start to hang

### PR DESCRIPTION
**Pull request by addisonElliott**
`Server.start` and `Mongos.start` functions search for the string "waiting for connections" to appear in the stdout of the spawned process. From there, it calls `waitForAvailable` to periodically try and connect to the server. 

These two functions will hang if the `--logpath` parameter is set because all of the output goes to a log file rather than stdout.

The solution proposed in this PR is to implement an initial delay in `waitForAvailable` to wait before attempting to connect. At that point, it continues the normal process of attempting to connect to the server at regular intervals for a specified number of times.

In addition, updated the `waitForAvailable` and `checkAvailable` functions to use *co* library rather than callbacks. Follows structure of rest of library well.

**Note:** Initial delay is set to 300ms and the retry delay is 100ms. The retry count is set to 25 because tests failed on a Linux machine for me while developing. May or may not be necessary since I've fixed other issues with the Travis CI build system.

### CODE REVIEW

**Comment by mbroadst** [comment]
@addisonElliott it looks like this needs to be rebased, and also there are a few obvious cleanups I think we can do based on previous discussion (e.g. no need to wrap `co` with a `Promise`)

File: lib/mongos.js [hunk 1]
```diff
@@ -317,6 +290,23 @@ class Mongos extends EventEmitter {
 
           self.state = 'stopped';
         });
+
+        // Periodically attempt to connect to server and return if successful
```
Comment (addisonElliott) [inline lib/mongos.js]: I was thinking about this part and I wanted your opinion.

With the old logic, Promise would resolve if data was printed to stdout saying "waiting for connections". There is also an error and close handler that will reject the Promise.

With the new logic, the Promise resolves if able to connect but the error and close handler are still there to reject the Promise.

Should we reject the Promise in the error/close handler now? Rather, would it be better just to print the error but not reject the Promise and let it timeout? Not sure if this is good practice.

File: lib/utils.js [hunk 2]
```diff
@@ -54,34 +56,39 @@ function checkAvailable(host, port, callback) {
...
-    });
-  }
+function waitForAvailable(host, port, options) {
+  return co(function*() {
+    options = Object.assign(
+      {},
+      {
+        initialMS: 300,
+        retryMS: 100,
+        retryCount: 25
```
Comment (addisonElliott) [inline lib/utils.js]: For the `retryCount`, what do you think is a good number for this?

The Linux errors I was receiving could have been unrelated which is why I originally had it at 100. With a `retryCount` of 100, that is a maximum wait time of 10.3s. At 25, the wait time is a maximum of 2.8s.

The bottleneck here is we want enough time for the server to get setup and throw errors. I don't have a lot of experience with MongoDB, but 2.8s seems like enough time as an average?

**Comment by addisonElliott** [comment]
@mbroadst This is unrelated to this PR, but it's a small comment.

I noticed that the [latest build for master failed](https://travis-ci.org/mongodb-js/mongodb-topology-manager/builds/471020370). It's the "address already in use" message again.

Would it work if we setup TravisCI to use kill-mongodb only? In other words, is there anyway that would work for you guys to **only** kill MongoDB instances when running tests on TravisCI?

File: lib/mongos.js (hunk 1 above)
Comment (mbroadst) [reply lib/mongos.js]: I would argue that you want to keep the handlers there to reject the promise, because we _do_ want to know if some error happens or if the binary unexpectedly quits. If you have a redundant rejection it will not cause any harm.

**Comment by mbroadst** [comment]
@addisonElliott I'd really like to stress that using `kill-mongodb` is the wrong solution to that problem. What we are experiencing there is almost certainly an edge case of this very module _not doing its job properly_.  I just ran the tests locally and experienced the same issue, but it seems somewhat random right now, so I'll take a look into what might be going on.

File: lib/mongos.js (hunk 1 above)
Comment (addisonElliott) [reply lib/mongos.js]: Okay, that's fine. Just wanted to bring the point up to discuss. Im in agreement then!

**Comment by addisonElliott** [comment]
Fair enough. I was under the impression when debugging that it was due to existing MongoDB instances running.

But, I was never able to reproduce the issue locally on Windows or Linux VM.

**Comment by addisonElliott** [comment]
@mbroadst when you get some free time, do you mind reviewing some of the PRs I've created please?

**Comment by mbroadst** [comment]
@addisonElliott yep, was on vacation. Tomorrow is a slower day at work, so I was going to budget some time then to review. Thanks for your patience!

File: test/single_server_tests.js [hunk 3]
```diff
@@ -187,5 +187,33 @@ describe('Server', function() {
...
+        managers.push(server);
+
+        // Purge the directory
+        yield server.purge();
+
+        // Start process
+        yield server.start();
+
+        // Stop the process
+        yield server.stop();
```
Comment (mbroadst) [inline test/single_server_tests.js]: it doesn't appear that anything is asserted in this test

File: lib/server.js [hunk 4]
```diff
@@ -415,6 +388,23 @@ class Server extends EventEmitter {
 
           self.state = 'stopped';
         });
+
+        // Periodically attempt to connect to server and return if successful
```
Comment (mbroadst) [inline lib/server.js]: Sorry to do this, but is it possible to separate this PR between listening for the known log messages (`'waiting for connections'`, etc), and refactoring `waitForAvailable` and `checkAvailable`? I think the changes to those functions are uncontroversial and could be merged immediately.

Removing checks for the log messages, however, is potentially a breaking behavioral change. I'm not quite sure what the implications are, so it behooves us to spend a little more time looking into the change.

**Review by mbroadst — CHANGES_REQUESTED** [review]

File: test/single_server_tests.js (hunk 3 above)
Comment (addisonElliott) [reply test/single_server_tests.js]: There isn't any assertion to be made. The issue that occurs in this test is that `server.start()` hangs forever because it's waiting for output in `stdout` when it's all going to the log.

Not sure what assertion should be made? Should I make a note about this?

File: lib/server.js (hunk 4 above)
Comment (addisonElliott) [reply lib/server.js]: It's no problem, I can do that.

I understand your concern in regard to searching for the 'waiting for connections'. However, this is the one breaking change that is stopping me from implementing `mongodb-topology-manager` in `mongodb-runner` because there is typically a `logpath` given.

Also, I think searching for the 'waiting for connections' message is a bit of a hack, but I suppose polling the server at regular intervals is a bit hacky as well.

**Comment by addisonElliott** [comment]
Once #25 is merged, I will rebase this PR and update it accordingly to only include the changes for logpath.

Am interested in hearing your thoughts on how we should go about fixing this issue? If we set the `initialMS` and `retryMS` correctly, it should be fine, but there is the concern that the server will take longer than the alloted time to start up. (This could be solved by increasing the parameter values except we don't allow specifying any options to the `start` function).

Otherwise, the only other approach I can think of is to regularly parse the logpath file for the string `waiting for connections` but I'm not sure I like that approach anymore.

**Comment by mbroadst** [comment]
Hi @addisonElliott, so sorry again for the delay! My teammate @daprahamian will be taking over review here so we can get some movement.

**Comment by addisonElliott** [comment]
@mbroadst It's no problem, I have a few other things on my plate so I haven't had a chance to edit the other PRs yet.

File: lib/utils.js (hunk 2 above)
Comment (daprahamian) [reply lib/utils.js]: I think 25 is fine.

**Review by daprahamian — APPROVED** [review]

**Review by mbroadst — APPROVED** [review]

**Comment by addisonElliott** [comment]
Sorry guys, haven't had a whole lot of time to work on this. Is there anything needed from me for this PR? I haven't had a chance to take a look at the other PRs in awhile...

**Comment by daprahamian** [comment]
@addisonElliott it looks like the CI has failed. is that b/c this needs #21 in order to pass?

**Comment by addisonElliott** [comment]
@daprahamian The "addr in use" message seems to be random and I cannot reproduce locally. I did fix one piece of code that **could** be causing the message, but I think the addr in use error still occurred after fixing that if I remember correctly (it's been awhile since I've looked at it).

Thus, if you were to rerun the CI, I think this PR should pass

**Comment by daprahamian** [comment]
Merged. Thank you so much @addisonElliott for your work on this!
